File: app/src/main/java/remix/myplayer/ui/fragment/LyricFragment.java
package remix.myplayer.ui.fragment;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import io.reactivex.disposables.Disposable;
import remix.myplayer.App;
import remix.myplayer.R;
import remix.myplayer.bean.mp3.Song;
import remix.myplayer.interfaces.OnInflateFinishListener;
import remix.myplayer.lyric.LrcView;
import remix.myplayer.lyric.SearchLrc;
import remix.myplayer.ui.fragment.base.BaseMusicFragment;
import remix.myplayer.util.LogUtil;
import remix.myplayer.util.SPUtil;
import remix.myplayer.util.ToastUtil;

/**
 * Created by Remix on 2015/12/2.
 */

/**
 * 歌词界面Fragment
 */
public class LyricFragment extends BaseMusicFragment {
    private OnInflateFinishListener mOnFindListener;
    private Song mInfo;
    @BindView(R.id.lrc_view)
    LrcView mLrcView;
    @BindView(R.id.offset_container)
    View mOffsetContainer;

    private Disposable mDisposable;

    public void setOnInflateFinishListener(OnInflateFinishListener l){
        mOnFindListener = l;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mPageName = LyricFragment.class.getSimpleName();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_lrc,container,false);

        mUnBinder = ButterKnife.bind(this,rootView);
        if(mOnFindListener != null)
            mOnFindListener.onViewInflateFinish(mLrcView);
        mInfo = getArguments().getParcelable("Song");

        updateLrc(mInfo);
        return rootView;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if(mDisposable != null){
            mDisposable.dispose();
        }
    }

    public void updateLrc(Song song){
        updateLrc(song,false);
    }

    public void updateLrc(Song song,boolean clearCache) {
        mInfo = song;
        getLrc("",clearCache);
    }

    public void updateLrc(String lrcPath){
        getLrc(lrcPath,true);
    }

    private void getLrc(String manualPath, boolean clearCache) {
        if (mInfo == null) {
            mLrcView.setText(getStringSafely(R.string.no_lrc));
            return;
        }
        if(clearCache){
            //清除offset
            SPUtil.putValue(App.getContext(),SPUtil.LYRIC_OFFSET_KEY.NAME,mInfo.getId() + "",0);
            if(mLrcView != null){
                mLrcView.setOffset(0);
            }
        }
        final int id = mInfo.getId();
        mDisposable = new SearchLrc(mInfo).getLyric(manualPath,clearCache)
                .doOnSubscribe(disposable -> mLrcView.setText(getStringSafely(R.string.searching)))
                .subscribe(lrcRows -> {
                    if (id == mInfo.getId() && mLrcView != null) {
                        if (lrcRows == null || lrcRows.size() == 0) {
                            mLrcView.setText(getStringSafely(R.string.no_lrc));
                            return;
                        }
                        mLrcView.setOffset(SPUtil.getValue(mContext,SPUtil.LYRIC_OFFSET_KEY.NAME,mInfo.getId() + "",0));
                        mLrcView.setLrcRows(lrcRows);
                    }
                }, throwable -> {
                    LogUtil.e(throwable);
                    if (id == mInfo.getId() && mLrcView != null) {
                        mLrcView.setLrcRows(null);
                        mLrcView.setText(getStringSafely(R.string.no_lrc));
                    }
                });
    }

    @OnClick({R.id.offset_reduce, R.id.offset_add,R.id.offset_reset})
    void onClick(View view){
        int original = SPUtil.getValue(mContext,SPUtil.LYRIC_OFFSET_KEY.NAME,mInfo.getId() + "",0);
        switch (view.getId()){
            case R.id.offset_reset:
                if(original != 0){
                    original = 0;
                    SPUtil.putValue(mContext,SPUtil.LYRIC_OFFSET_KEY.NAME,mInfo.getId() + "",0);
                }
                ToastUtil.show(mContext,R.string.lyric_offset_reset);
                break;
            case R.id.offset_add:
                original += 500;
                SPUtil.putValue(mContext,SPUtil.LYRIC_OFFSET_KEY.NAME,mInfo.getId() + "",original);
                break;
            case R.id.offset_reduce:
                original -= 500;
                SPUtil.putValue(mContext,SPUtil.LYRIC_OFFSET_KEY.NAME,mInfo.getId() + "",original);
                break;
        }
        if(original != 0){
            ToastUtil.show(mContext,original > 0 ? R.string.lyric_advance_x_second : R.string.lyric_delay_x_second,
                    String.format(Locale.getDefault(),"%.1f",original / 1000f));
        }
        mLrcView.setOffset(original);
    }

    public void showLyricOffsetView(){
        mOffsetContainer.setVisibility(View.VISIBLE);
        mOffsetContainer.postDelayed(() -> mOffsetContainer.setVisibility(View.GONE),3000);
    }
}

File: app/src/main/java/remix/myplayer/ui/fragment/LyricFragment.kt
package remix.myplayer.ui.fragment

import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.os.Message
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Consumer
import kotlinx.android.synthetic.main.fragment_lrc.*
import remix.myplayer.App
import remix.myplayer.R
import remix.myplayer.bean.mp3.Song
import remix.myplayer.helper.MusicServiceRemote
import remix.myplayer.lyric.LyricSearcher
import remix.myplayer.misc.handler.MsgHandler
import remix.myplayer.misc.handler.OnHandleMessage
import remix.myplayer.misc.interfaces.OnInflateFinishListener
import remix.myplayer.theme.ThemeStore
import remix.myplayer.ui.fragment.base.BaseMusicFragment
import remix.myplayer.util.SPUtil
import remix.myplayer.util.ToastUtil
import timber.log.Timber
import java.util.*
import kotlin.math.abs

/**
 * Created by Remix on 2015/12/2.
 */

/**
 * 歌词界面Fragment
 */
class LyricFragment : BaseMusicFragment(), View.OnClickListener {
  private var onFindListener: OnInflateFinishListener? = null
  private var song: Song? = null

  private var disposable: Disposable? = null
  private val msgHandler = MsgHandler(this)

  private val lyricSearcher = LyricSearcher()

  fun setOnInflateFinishListener(l: OnInflateFinishListener) {
    onFindListener = l
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    pageName = LyricFragment::class.java.simpleName
  }

  override fun onCreateView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View? {
    return inflater.inflate(R.layout.fragment_lrc, container, false)
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)

    offsetReduce.setOnClickListener(this)
    offsetAdd.setOnClickListener(this)
    offsetReset.setOnClickListener(this)
    onFindListener?.onViewInflateFinish(lrcView)
    //黑色主题着色按钮
    val themeRes = ThemeStore.themeRes
    if (themeRes == R.style.Theme_APlayer_Black || themeRes == R.style.Theme_APlayer_Dark) {
      iv_offset_reduce_arrow.setColorFilter(Color.WHITE)
      iv_offset_reduce_second.setColorFilter(Color.WHITE)
      offsetReset.setColorFilter(Color.WHITE)
      iv_offset_add_arrow.setColorFilter(Color.WHITE)
      iv_offset_add_second.setColorFilter(Color.WHITE)
    }
  }

  override fun onDestroyView() {
    msgHandler.remove()
    disposable?.dispose()
    onFindListener = null
    super.onDestroyView()
  }

  @JvmOverloads
  fun updateLrc(song: Song, clearCache: Boolean = false) {
    this.song = song
    getLrc(Uri.EMPTY, clearCache)
  }

  fun updateLrc(uri: Uri) {
    getLrc(uri, true)
  }

  private fun getLrc(uri: Uri, clearCache: Boolean) {
    if (!isVisible)
      return
    if (song == null) {
      lrcView.setText(getStringSafely(R.string.no_lrc))
      return
    }
    if (clearCache) {
      //清除offset
      SPUtil.putValue(App.context, SPUtil.LYRIC_OFFSET_KEY.NAME, song?.id.toString() + "", 0)
      lrcView.setOffset(0)
    }
    val id = song?.id

    disposable?.dispose()
    Timber.v("setSearching")
    lrcView.setText(getStringSafely(R.string.searching))
    disposable = lyricSearcher.setSong(song ?: return)
      .getLyricObservable(uri, clearCache)
      .subscribe(Consumer {
        Timber.v("setLrcRows")
        if (id == song?.id) {
          if (it == null || it.isEmpty()) {
            lrcView.setText(getStringSafely(R.string.no_lrc))
            return@Consumer
          }
          lrcView.setOffset(
            SPUtil.getValue(
              requireContext(),
              SPUtil.LYRIC_OFFSET_KEY.NAME,
              song?.id.toString() + "",
              0
            )
          )
          lrcView.setLrcRows(it)
        }
      }, Consumer {
        Timber.v(it)
        if (id == song?.id) {
          lrcView.setLrcRows(null)
          lrcView.setText(getStringSafely(R.string.no_lrc))
        }
      })
  }

  override fun onClick(view: View) {
    msgHandler.removeMessages(MESSAGE_HIDE)
    msgHandler.sendEmptyMessageDelayed(MESSAGE_HIDE, DELAY_HIDE)

    val originalOffset =
      SPUtil.getValue(requireContext(), SPUtil.LYRIC_OFFSET_KEY.NAME, song?.id.toString() + "", 0)
    var newOffset = originalOffset
    when (view.id) {
      R.id.offsetReset -> {
        newOffset = 0
        ToastUtil.show(requireContext(), R.string.lyric_offset_reset)
      }
      R.id.offsetAdd -> newOffset += 500
      R.id.offsetReduce -> newOffset -= 500
    }
    if (originalOffset != newOffset) {
      SPUtil.putValue(
        requireContext(),
        SPUtil.LYRIC_OFFSET_KEY.NAME,
        song?.id.toString() + "",
        newOffset
      )
      val toastMsg = msgHandler.obtainMessage(MESSAGE_SHOW_TOAST)
      toastMsg.arg1 = newOffset
      msgHandler.removeMessages(MESSAGE_SHOW_TOAST)
      msgHandler.sendMessageDelayed(toastMsg, DELAY_SHOW_TOAST)
      lrcView.setOffset(newOffset)
      MusicServiceRemote.setLyricOffset(newOffset)
    }

  }

  fun showLyricOffsetView() {
    if (lrcView.getLrcRows() == null || lrcView.getLrcRows()?.isEmpty() == true) {
      ToastUtil.show(requireContext(), R.string.no_lrc)
      return
    }
    offsetContainer.visibility = View.VISIBLE
    msgHandler.sendEmptyMessageDelayed(MESSAGE_HIDE, DELAY_HIDE)
  }

  fun setLyricScalingFactor(choose: Int) {
    val factor = when (choose) {
      0 -> {
        1f
      }
      1 -> {
        1.5f
      }
      2 -> {
        2f
      }
      else -> {
        1f
      }
    }
    lrcView.setLrcScalingFactor(factor)
  }

  @OnHandleMessage
  fun handleInternal(msg: Message) {
    when (msg.what) {
      MESSAGE_HIDE -> {
        offsetContainer.visibility = View.GONE
      }
      MESSAGE_SHOW_TOAST -> {
        val newOffset = msg.arg1
        if (newOffset != 0 && abs(newOffset) <= 60000) {//最大偏移60s
          ToastUtil.show(
            requireContext(),
            if (newOffset > 0) R.string.lyric_advance_x_second else R.string.lyric_delay_x_second,
            String.format(Locale.getDefault(), "%.1f", newOffset / 1000f)
          )
        }
      }
    }
  }


  companion object {
    private const val DELAY_HIDE = 5000L
    private const val DELAY_SHOW_TOAST = 100L

    private const val MESSAGE_HIDE = 1
    private const val MESSAGE_SHOW_TOAST = 2
  }

}
